src/library_wasmfs_opfs.js

```diff
@@ -0,0 +1,250 @@
+/**
+ * @license
+ * Copyright 2022 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+mergeInto(LibraryManager.library, {
+  $wasmfsOPFSDirectories: {
+    allocated: [],
+    free: [],
+    get: function(i) {
+      return this.allocated[i];
+    }
+  },
+
+  $wasmfsOPFSFiles: {
+    allocated: [],
+    free: [],
+    get: function(i) {
+      return this.allocated[i];
+    }
+  },
+
+  $wasmfsOPFSAccesses: {
+    allocated: [],
+    free: [],
+    get: function(i) {
+      return this.allocated[i];
+    }
+  },
+
+  $wasmfsOPFSAllocate: function(ids, handle) {
+    let id;
+    if (ids.free.length > 0) {
+      id = ids.free.pop();
+      ids.allocated[id] = handle;
+    } else {
+      id = ids.allocated.length;
+      ids.allocated.push(handle);
+    }
+    return id;
+  },
+
+  $wasmfsOPFSFree: function(ids, id) {
+    delete ids.allocated[id];
+    ids.free.push(id);
+  },
+
+  _wasmfs_opfs_init_root_directory__deps: ['$wasmfsOPFSDirectories'],
+  _wasmfs_opfs_init_root_directory: async function(ctx) {
+    if (wasmfsOPFSDirectories.allocated.length == 0) {
+      // Directory 0 is reserved as the root
+      let root = await navigator.storage.getDirectory();
```

Comment:
FYI, this is the only line that is specific to the _Origin Private_ part of the File System Access API. This interface could be made more general by allowing the user to provide a directory handle (instead of calling the OPFS version of it). They could choose to give the origin private directory, or use `window.showDirectoryPicker` to prompt the user for a real folder.

It could be good to keep using `navigator.storage.getDirectory()` as a default, though.

Reply:
Oh, this is wrong, sorry. I hadn't realized that the sync handle is limited to the [OPFS](https://github.com/WICG/file-system-access/blob/main/AccessHandle.md#non-goals). That's unfortunate.

Any plans on providing a backend for the normal FS API too, so files can be backed by real folders on disk?

Reply:
I don't have concrete plans to work on that myself, but it would certainly be nice to have. My hope is that we can make it easy to create new WasmFS backends as userspace JS libraries, but it would also be reasonable to generalize this backend once it lands.

Reply:
Ah, interesting. I'll keep an eye on this space and would be happy to contribute such a library when possible.

Reply:
The ability to create and mount custom backends in WasmFS is awesome.

I've been experimenting with window.showDirectoryPicker.  The OPFS backend does most of the work already, but I found several challenges when replacing the root handle with a native `FileSystemDirectoryHandle`.

- The lack of sync handles requires using createWritable instead.  The OPFS backend already uses createWritable when pthreads are not available, so we just need to always include the FileSystemAsyncAccessHandle class even when pthreads are available and avoid creating sync handles in the native access case.  This requires the most changes (mostly minor).
- `SharedArrayBuffer` cannot be used, so the data must be copied to a temporary `ArrayBuffer`.
- Passing the native `FileSystemDirectoryHandle` into the backend is difficult.  I had to use IndexedDB to store it, then retrieve it when creating the backend.  Is there a better way?
- If the handle is persisted across sessions, some way of calling `requestPermission` on the handle is required from the backend (when `getEntries` is called).  This is asynchronous and cannot be done from the proxy thread.

Working around these challenges in a custom backend based on OPSF backend code, I was able to view, create, delete files and directories on the native OS.  I think the OPFS backend could be more generalized with a `native` option so that a whole new backend does not need to be created for native file access.  The question is how to specify the `FileSystemDirectoryHandle` when creating the backend.  Using IndexedDB adds some overhead and complexity.  In my app, this works fine because I need to persist it anyway.

Some other issues I noticed include:
in wasmfs.h:
 - `wasmfs_unmount` takes an `intptr_t` instead of a `const char *`
 - `wasmfs_get_backend_by_path` takes a `char *` instead of a `const char *`

in opfs_backend.cpp:
 - `~OPFSDirectory` compares the `dirID` to zero to avoid freeing the root ID, however isn't the root ID 1 (zero is undefined)?  I think there may have been a couple of other places where ID 0 is assumed to be the root.

Reply:
Great feedback, thanks @goldwaving.

 - Adding a `native` option to the OPFS backend code to reduce deduplication sounds like a good idea to me.
 - To pass the `FileSystemDirectoryHandle` to the backend, I hope it would be possible to just store it in a global variable or a fancier global registry so the backend can retrieve it. This might also be an interesting use case for clang's [`externref` support](https://github.com/llvm/llvm-project/blob/ba1349fc31295a3670b34c189838a133e18c0bed/clang/test/CodeGen/WebAssembly/wasm-externref.c): we could pass the handle directly through C to the backend constructor.
 - The assumption that the root has ID 0 sounds like a bug. I think that used to be true, but we must not have updated all the code when we changed that.

Reply:
A global variable would work for single threaded apps.  When using pthreads, however, the FileSystemDirectoryHandle may have to cross two or more thread barriers (from main, to app's proxy, then to OPFS's proxy).  That's where it gets tricky.  Is there an emscripten way to move the handle across threads or is IndexedDB the only option?

Reply:
Oh I see, that does sound tricky. @sbc100, do you know if we provide a way for users to postMessage JS objects from one thread to another?

Reply:
Not really.. thats kind of below the abstraction level that our APIs provide.

Many folks have asked for this over the years but we have normally managed to stear them away from needing it at all.

In theory is possible to this this today by looking up wither worker object for a given thread and just using `postMessage` directly.   You would also need to incercept/override the message handler for the thread, but that should also be doable today.  

We could add a dedicated API for this, but I'd be tempted to just expose and official pthread->worker lookup function and then show an example of how to use postMessage based on that.